### ISSUE TITLE
memory leak in LinearNDInterpolator

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#9747]
Here is a reported memory leak in `interpolate.LinearNDInterpolator`: https://mail.python.org/pipermail/scipy-dev/2018-November/023201.html

Setting `NPY_RELAXED_STRIDES_CHECKING=0` seems to remove the memory leak (message at end of that email thread). This needs further investigation.

**Comment by pv** [comment]
Cython bug is a possibility

**Comment by AtsushiSakai** [comment]
This issue still exists in SciPy v1.5.1..

Script output:
```
rss_used=         0 KB vms_used=         0 KB
rss_used=      2668 KB vms_used=      2992 KB
rss_used=      5904 KB vms_used=      6116 KB
rss_used=      8568 KB vms_used=      7940 KB
rss_used=     11488 KB vms_used=     11140 KB
rss_used=     15532 KB vms_used=     15224 KB
rss_used=     19924 KB vms_used=     19968 KB
rss_used=     23624 KB vms_used=     22652 KB
rss_used=     27460 KB vms_used=     27784 KB
rss_used=     31432 KB vms_used=     31796 KB
rss_used=     35572 KB vms_used=     36248 KB (final)
```

My environment is 
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
# 1.5.1 1.19.0 sys.version_info(major=3, minor=8, micro=2, releaselevel='final', serial=0)
```

**Comment by peterbell10** [comment]
The culprit seems to be this line:
https://github.com/scipy/scipy/blob/283cdbb3fc48a4a874f14b13c25b4c11ad543cb5/scipy/interpolate/interpnd.pyx#L258

I can reproduce the leak with the entire `_do_evaluate` function reduced to that line, and commenting out that line fixes the leak. So it looks like a cython bug.

Probably the root cause is that `self.values` has a zero-strided axis. `self.values.strides == (8, 0)`. A simple workaround is to add `self.values = self.values.copy()` to the constructor.

**Comment by peterbell10** [comment]
Minimised and reported upstream as cython/cython#3747.

**Comment by AtsushiSakai** [comment]
@peterbell10 Do you think #12499 is related the cython bug which you reported?

When I run this code based on your issue:
```
%%cython
def memory_leak(a):
    cdef double [:,::1] arr = a
    return

a = np.random.rand(10)[:, None]
a.setflags(write=False)
memory_leak(a)
``` 

I got a same error message reported in #12499 
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-21-a70a6a01379b> in <module>
      1 a = np.random.rand(10)[:, None]
      2 a.setflags(write=False)
----> 3 memory_leak(a)

_cython_magic_bc07bf665e37a3b799220af35bd6f6da.pyx in _cython_magic_bc07bf665e37a3b799220af35bd6f6da.memory_leak()

~/.ipython/cython/_cython_magic_bc07bf665e37a3b799220af35bd6f6da.cpython-38-darwin.so in View.MemoryView.memoryview_cwrapper()

~/.ipython/cython/_cython_magic_bc07bf665e37a3b799220af35bd6f6da.cpython-38-darwin.so in View.MemoryView.memoryview.__cinit__()

ValueError: buffer source array is read-only
```

**Comment by peterbell10** [comment]
No, I don't think related. `double [:, :]` is a read-write view so cython is right to reject read-only arrays. The fix there is to add `const` to any inputs that aren't modified.
Ref: https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#read-only-views

**Comment by AtsushiSakai** [comment]
Oh. OK. Thanks!!

**Comment by da-woods** [comment]
For what it's worth I think the bug's in Numpy rather than Cython: https://github.com/cython/cython/issues/3747#issuecomment-662870252

**Comment by peterbell10** [comment]
Thanks @da-woods, reported to NumPy in numpy/numpy#16934

**Comment by ev-br** [comment]
Note : apparently, the upstream fix is going to ve in numpy 1.20 https://github.com/numpy/numpy/issues/16934

**Comment by AtsushiSakai** [comment]
I confirmed this issue was solved using numpy 1.20.

```
1.6.0 1.20.0 sys.version_info(major=3, minor=9, micro=1, releaselevel='final', serial=0)
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB
rss_used=         0 KB vms_used=         0 KB (final)
```
